### PR TITLE
Upgrade extension to JupyterLab 4

### LINKED ISSUE #116: Make the extension compatible for JupyterLab 4
JupyterLab v4 was released on May 15th: https://github.com/jupyterlab/jupyterlab/releases/tag/v4.0.0

I tested with release candidates before and it seems that there are issues arising in the back-end (front-end looks fine). This is perhaps due to updates in `jupyter-server`.

### PR DESCRIPTION
**Pull request by mahendrapaipuri**
Hello,

This PR focuses on upgrading build, run time dependencies to JupyterLab 4. Following changes are made:

- Build time dep is bumped to JupyterLab 4 and Jupyter Server 2
- Min Python version is bumped to `3.8` inline with JupyterLab 4
- Node deps are bumped to JupyterLab 4, Lumino 2 and React 18
- Minor changes are made in frontend extension to account for changes in JupyterLab 4
- CI build workflow now uses [jupyterlab maintainer tools](https://github.com/jupyterlab/maintainer-tools/tree/main) to setup build env

Closes #116

### CODE REVIEW

**Comment by loichuder** [comment]
Wow, thanks for the great PR :heart: It looks great !

Unfortunately, there appear to be some peer deps issues in the build of the CI :disappointed:  It would great if you can take a look to see if you can reproduce it. 

In any case, I will test it myself later this week.

**Comment by mahendrapaipuri** [comment]
@loichuder Cheers, happy to contribute. I tried to fix all the potential issues in CI. Should be good to go.

File: README.md [hunk 1]
```diff
@@ -15,11 +15,14 @@ See [Usage](README.md#Usage) for more details.
 
 ## Requirements
 
-- JupyterLab 3
+- JupyterLab >= 4
```
Comment (loichuder) [inline README.md:18]: With JupyterLab 5 surely coming, I think it is better to drop the `>=` here :smile:  
```suggestion
- JupyterLab 4
```

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "jupyterlab-h5web",
-  "version": "8.0.0",
+  "version": "8.1.0",
```
Comment (loichuder) [inline package.json:3]: You can leave the version number unchanged. 

We have a dedicated release process that update it automatically.

File: setup.py [hunk 3]
```diff
@@ -54,24 +54,25 @@
...
-    install_requires=["jupyter_server>=1.6,<2", "h5grove==1.3.0", "h5py>=3.5"],
+    install_requires=["jupyter_server>=1.6,<3", "h5grove==1.3.0", "h5py>=3.5"],
     extras_require={"full": ["hdf5plugin"]},
-    python_requires=">=3.7",
+    python_requires=">=3.8",
     zip_safe=False,
     include_package_data=True,
     platforms="Linux, Mac OS X, Windows",
-    keywords=["Jupyter", "JupyterLab", "JupyterLab3"],
+    keywords=["Jupyter", "JupyterLab", "JupyterLab3", "JupyterLab4"],
```
Comment (loichuder) [inline setup.py:63]: I didn't check but I guess the extension is not back-compatible with JupyterLab3, right ?

If so, I wonder if we should not drop the JLab3 `keywords` and `classifiers`

File: src/browser.ts [hunk 4]
```diff
@@ -51,9 +51,9 @@ export function activateOpenInBrowser(
     execute: () => {
       const file = browser.selectedItems().next();
```
Comment (loichuder) [inline src/browser.ts:52]: I cannot open the HDF5 file by double-clicking in the browser :disappointed: ...

It tells me that `browser` is `undefined` here. Is it working fine for you ?

**Review by loichuder — CHANGES_REQUESTED** [review]
Amazing, thanks for the hard work!

Just a few changes related to how we manage the project. I also cannot open the HDF5 files in the browser for some reason :confused:

File: setup.py (hunk 3 above)
Comment (mahendrapaipuri) [reply setup.py:63]: That is a very good point and seems like folks at [Jupyter](https://github.com/jupyterlab/jupyter-renderers/blob/4d08ef19cd90d56c121587f62764ee2a135caa98/packages/fasta-extension/pyproject.toml#L13) agree with you. We should drop JupyterLab3 from keywords and classifiers

File: src/browser.ts (hunk 4 above)
Comment (mahendrapaipuri) [reply src/browser.ts:52]: Are you sure that you are on JupyterLab 4? I guess you are using JupyterLab 3 and that's where the `browser` is `undefined` error comes from.

File: src/fileType.ts [hunk 5]
```diff
@@ -7,13 +7,13 @@ import { hdf5Icon } from './icons';
...
-const HDF5_FILE_TYPE: Partial<DocumentRegistry.IFileType> = {
+const HDF5_FILE_TYPE = {
   name: 'hdf5file',
   icon: hdf5Icon,
   displayName: 'HDF5 File',
   extensions: ['.cxi', '.hdf', '.hdf5', '.h5', '.nexus', '.nx', '.nx5', '.nxs'],
   mimeTypes: ['application/x-hdf5'],
   fileFormat: 'base64' as const,
-};
+} satisfies Partial<DocumentRegistry.IFileType>;
```
[the remark this answers is not in the document]
Comment (mahendrapaipuri) [reply src/fileType.ts:17]: Didnt know about `satisfies` operator. Cheers for the pointer!!

File: src/browser.ts (hunk 4 above)
Comment (loichuder) [reply src/browser.ts:52]: Pretty sure I was using JLab4. I'll have a second go.

File: README.md (hunk 1 above)
Comment (mahendrapaipuri) [reply README.md:18]: Fair enough, I agree!!

File: src/browser.ts (hunk 4 above)
Comment (mahendrapaipuri) [reply src/browser.ts:52]: Screenshot_20230628_104103

I started in a clean env and I could open the files.

Comment (loichuder) [reply src/browser.ts:52]: Yeah, you were right: the problem was on my end. I was indeed on JLab4 but had probably some stale builds lying around.

I could make it work locally and on Binder :+1:

**Comment by mahendrapaipuri** [comment]
Btw, [hatch](https://hatch.pypa.io/latest/) is the recommended build system by project Jupyter to build extensions and they have created a [plugin](https://github.com/jupyterlab/hatch-jupyter-builder) for building extensions. Here is a [reference](https://github.com/jupyterlab/jupyter-renderers/blob/master/packages/fasta-extension/pyproject.toml).

Are you keen in migrating to this build system? It is pretty neat and avoid custom logic in `setup.py`.

**Comment by loichuder** [comment]
> Btw, [hatch](https://hatch.pypa.io/latest/) is the recommended build system by project Jupyter to build extensions and they have created a [plugin](https://github.com/jupyterlab/hatch-jupyter-builder) for building extensions. Here is a [reference](https://github.com/jupyterlab/jupyter-renderers/blob/master/packages/fasta-extension/pyproject.toml).
> 
> Are you keen in migrating to this build system? It is pretty neat and avoid custom logic in `setup.py`.

`hatch` looks indeed seducing. I had a quick experience with it in another project: I encountered some problems and could not find much support online. I guess the project is still a bit young ?

Personally, I would prefer to wait a bit to see how `hatch` fares in the near future. But if you think you can make it work, then I'll welcome another PR.

**Comment by mahendrapaipuri** [comment]
Yes, it is quite a young project and I had difficulties with it too. I had to dig into source code as there is not a lot of support.

I have already done it for quite some of our custom extensions. I will put up a PR when I have some time and then you can choose to look into it when you are more comfortable working with it!!

**Review by loichuder — APPROVED** [review]
Thanks a lot for this ! 

This saved me a lot of time and energy.

**Pull request merged**